lib/rules/no-reassign-constant.js

```diff
@@ -0,0 +1,141 @@
+        var isWrite = (ref.isWrite() || !ref.isReadOnly());
+
+        if (!isWrite && isDeleted(ref.identifier.parent)) {
+            isWrite = true;
+        }
+
+        return isWrite;
+    }
+
+    /**
+     * Report a reference as invalid
+     *
+     * @param {String} name - the variables name
+     * @param {Reference} ref - the reference to report
+     * @returns {void}
+     * @private
+    */
+    function report(name, ref) {
+        context.report(ref.identifier, MESSAGE, {name: name});
+    }
+
+    /**
+     * Report a reference as invalid
+     *
+     * @param {Scope} scope - an escope object
+     * @param {Variable} variable - variable to apply references too
+     * @returns {Variable} returns the variable with references
+     * @private
+    */
+    function transformGlobalVariables(scope, variable) {
+        scope.references.forEach(function(ref) {
+            if (ref.identifier.name === variable.name) {
+                variable.references.push(ref);
+            }
+        });
+
+        return variable;
+    }
+
+    /**
+     * Checks for and reports reassigned constants
+     *
+     * @param {Scope} scope - an escope Scope object
+     * @returns {void}
+     * @private
+    */
+    function checkScope(scope) {
+        var variables = scope.variables;
+
+        if (!scope.functionExpressionScope) {
+            for (var i = 0, l = variables.length; i < l; ++i) {
+                // skip function expression names
+                if (variables[i].eslintUsed) {
+                    continue;
+                }
+
+                // skip implicit "arguments" variable
+                if (scope.type === "function" && variables[i].name === "arguments" && variables[i].identifiers.length === 0) {
+                    continue;
+                }
+
+                var def = variables[i].defs[0];
+
+                // only constants
+                if (!def || def.kind !== "const") {
+                    continue;
+                }
+
+                var variable = variables[i];
+                var references = variable.references;
+                var name = variable.name;
+                var assignments = references.filter(isReassignment);
+
+                if (assignments.length > 1) {
+                    // remove original assigment
+                    assignments.shift();
+
+                    assignments.forEach(report.bind(null, name));
+                }
+            }
+        }
+
+        scope.childScopes.forEach(checkScope);
+    }
+
+    return {
+        "Program:exit": function() {
+            var scope = context.getScope();
+
+            if (scope.type === "global") {
+                scope = {
+                    childScopes: scope.childScopes,
+                    variables: scope.variables.map(transformGlobalVariables.bind(null, scope))
+                };
+            }
+
+            if (context.ecmaFeatures.blockBindings) {
+                checkScope(scope);
```

Comment:
I'm not sure this is the best approach. Recursively searching through every scope is pretty inefficient. What you could do instead is listen for each `AssignmentExpression` and check to see if the lefthand side is a constant. That would dramatically cut down on the amount of work you're working doing for each file.